include/xrpl/protocol/Asset.h

```diff
@@ -26,10 +26,17 @@
 
 namespace ripple {
 
+class Asset;
+
 template <typename TIss>
 concept ValidIssueType =
     std::is_same_v<TIss, Issue> || std::is_same_v<TIss, MPTIssue>;
 
+template <typename A>
+concept AssetType =
+    std::is_convertible_v<A, Asset> || std::is_convertible_v<A, Issue> ||
+    std::is_convertible_v<A, MPTIssue> || std::is_convertible_v<A, MPTID>;
```

Comment:
are you sure you wanted `is_convertible_v` here ? You are basically telling the compiler to use anything that these types have converting constructors for; in which case `std::is_convertible_v<A, Asset>` should be enough because it has converting constructors from the other three.

Oh and `MPTID`, being a type alias to `base_uint` also has a converting constructor from `beast::Zero`, is the intent to also include this type ?

Reply:
You are right. Can just have this:

    std::is_convertible_v<A, Asset> || std::is_convertible_v<A, Issue>

Need `Issue` to handle XRP to Issue in unit-tests.

I think `beast::Zero` is fine.

Reply:
how about

```
std::is_same_v<std::remove_cvref_t<A>, Asset> || std::is_same_v<std::remove_cvref_t<A>, Issue> || std::is_same_v<std::remove_cvref_t<A>, MPTIssue> || std::is_same_v<std::remove_cvref_t<A>, MPTID>;
```
... or do you need converting constructors as well , in which case this concept would be better named `IsConvertibleToAsset` ?

?

Reply:
`IsConvertibleToAsset` sounds good.

Reply:
But, it's not just `Asset` though.

Reply:
oh wait I see. The original was

```
template <typename A>
concept AssetType =
    std::is_same_v<A, Asset> || std::is_convertible_v<A, Issue> ||
    std::is_convertible_v<A, MPTIssue> || std::is_convertible_v<A, MPTID>;
```

Perhaps keep it this way ?

Reply:
John suggested to change. I have to dig through the discussion threads to find out why. I'm fine with reverting.

Reply:
@Bronek try changing it in the `vault` branch and see if it builds. If it doesn't, that will be the reason I asked for the change. If it does, then the cause might have been an experiment that I removed.

Reply:
Thanks @thejohnfreeman I see it now. The `is_convertible_to<Asset>` enables the `STAmount` to take first parameter of any type convertible to `Asset`, including `STObject::ValueProxy<STIssue>` , which is returned from `STObject::at`.

If we change `is_convertible_to<Asset>` into `is_same_v<Asset>` we are getting compilation error in `View.cpp` (your `vault` branch) here:

```c++
[[nodiscard]] STAmount
sharesToAssetsWithdraw(
    ReadView const& view,
    std::shared_ptr<SLE> const& vault,
    STAmount const& shares)
{
    assert(shares.asset() == vault->at(sfMPTokenIssuanceID));
    Number assetTotal = vault->at(sfAssetTotal);
    assetTotal -= vault->at(sfLossUnrealized);
    STAmount assets{vault->at(sfAsset)};
    ^^^^^ No matching constructor for initialization of 'STAmount'
```

Even though the workaround is simple:

```c++
    Asset asset = vault->at(sfAsset);
    STAmount assets{asset};
```

... I think it's better if we do not have to use it. The name "proxy" implies that this needs to work without the explicit construction, after all.

@gregtatcam no change requested, this is good.

Reply:
Similar thing happens if we get rid of `std::is_convertible_v<A, MPTID>`, for similar reason

```c++
[[nodiscard]] STAmount
assetsToSharesWithdraw(
    ReadView const& view,
    std::shared_ptr<SLE> const& vault,
    STAmount const& assets)
{
    assert(assets.asset() == vault->at(sfAsset));
    Number assetTotal = vault->at(sfAssetTotal);
    assetTotal -= vault->at(sfLossUnrealized);
    STAmount shares{vault->at(sfMPTokenIssuanceID)};
    ^^^^^ No matching constructor for initialization of 'STAmount'
```

So I think we should keep this as-is, no change requested here @gregtatcam